### ISSUE TITLE
MNIST example uses DataModule setup hook wrong

### ISSUE THREAD

**Issue by AlexHarn** [ashleve/lightning-hydra-template#184]
First of all, I am sorry but I don't have the time right now to create a pull request for this. I might do so later, but I still wanted to raise awareness right now because I feel it is an important issue with the example, especially when people use that as a guideline of how to write their own project, like I did.

Here is the problem: When calling `trainer.test()` after `trainer.fit()`, `LightningDataModule.setup()` is called twice, which is especially problematic when using `random_split` in `setup()` because it leads to training samples leaking into the test set (in the worst case of extremely large training data compared to test data, the test data will most likely consist exclusively of training samples).

At first I thought this was a PyTorchLightning bug, but it apparently is not. See the issue I created for reference [here](https://github.com/PyTorchLightning/pytorch-lightning/issues/9865). There is also a potential fix proposed there that could be applied for this template: making the `setup` method idempotent.

**Comment by ashleve** [comment]
Hi @AlexHarn, thanks for making this issue! 
Must admit I was not aware of this behavior.

One more thing to realize here is that setting seed in `train.py` makes the random split always the same, so I believe there should be no samples leaking to test set as long as the seed is specified in config.

Thanks again and you don't need to bother with the PR, I will probably make the `setup` idempotent as you suggest
